dfc/callstats.go

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+ */
+
+package dfc
+
+// Keep track average latency of call between proxies and targets
+// Currently it filters the calls by request (configurable), other filters (for example,
+// by http method) can be added if need to
+// A warning is logged if a call's latency exceeds the average by a predefined factor/threshhold
+// APIs provided by this package are concurrent access safe
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type (
+	callInfo struct {
+		url     string
+		latency time.Duration
+		failed  bool
+	}
+
+	latency struct {
+		cnt   int64
+		total time.Duration // accumulated latency
+	}
+
+	// CallStatsServer stores parameters and call stats collected from each call
+	CallStatsServer struct {
+		wg              sync.WaitGroup
+		ch              chan callInfo
+		stats           map[string]*latency
+		factor          float32
+		requestIncluded []string
+	}
+)
+
+func (l *latency) avg() int64 {
+	if l.cnt == 0 {
+		return 0
+	}
+
+	return int64(l.total) / l.cnt
+}
+
+// NewCallStatsServer returns a CallStatsServer
+func NewCallStatsServer(requestsIncluded []string, factor float32) *CallStatsServer {
+	return &CallStatsServer{
+		ch:              make(chan callInfo, 100),
```

Comment:
Why 100?

Reply:
just a number I picked, any suggestions?

Reply:
Not sure what the optimal number here is - too low and it'll starting blocking. I would suggest to make it a const and give it a contextual comment so that it's easy to modify in the future if necessary.

Reply:
const is not necessary, comment I will add inside call stats.go, good?

Reply:
Ok!

Reply:
to me, worker function looks very fast, so I doubt 100 is not enough. E.g, keepalive happens once in a few tens of second

Reply:
I was thinking to make it in config file, but the config file is already huge